Replace squish with bcdec for BC decompression

BlueCube3310 opened the pull request:
Fixes #95721

Replaces libsquish with the bcdec library for decompressing BC-compressed textures. This improves decompression speeds (about 2x faster) and allows runtime decompression of BC6 and BC7. 
Some more comparisons can be found here: https://aras-p.info/blog/2022/06/23/Comparing-BCn-texture-decoders/ 
(BC6 is probably faster here since in the blog's benchmarks it's also converted from half to float, in here it's directly to half).

Performance:
image: https://polyhaven.com/a/worn_brick_floor (no mipmaps)
Compressed with DirectXTex
bc1/bc2/bc3/bc6/bc7 - diffuse 4k
bc4 - displacement 4k
bc5 - normal (openGL) 4k

OS: Windows 10
CPU: AMD Ryzen 9 5900X 12-Core Processor (24 threads)
RAM: 64 GB DDR4
build: production

| format | squish | cvtt | bcdec |
|---------|-------|------|-------|
| bc1 | 52ms | n/a | 24ms |
| bc2 | 58ms | n/a | 31ms |
| bc3 | 61ms | n/a | 35ms |
| bc4 | 48ms | n/a | 12ms |
| bc5 | 60ms | n/a | 28ms |
| bc6 | n/a | 364ms | 122ms |
| bc7 | n/a | 174ms | 133ms |

TODO:
- [x] Add a testing project,
- [x] Compare performance

MRP: [bcdectest.zip](https://github.com/user-attachments/files/17130853/bcdectest.zip)

BlueCube3310 committed:
Replace squish with bcdec for BC decompression

fire commented:
As someone who touched the older areas of the compression I think this is a good idea, but we probably want to merge code changing the 3d compression types earlier or push it for Godot Engine 4.5.

clayjohn commented:
> As someone who touched the older areas of the compression I think this is a good idea, but we probably want to merge code changing the 3d compression types earlier or push it for Godot Engine 4.5.

Keep in mind, this is only for _decompression_, so the impact will be much smaller than if we were to change the compression library

fire commented on modules/bcdec/image_decompress_bcdec.cpp:
```diff
@@ -0,0 +1,180 @@
+	const uint8_t *rb = p_image->get_data().ptr();
+	uint8_t *wb = data.ptrw();
+
+	// Decompress mipmaps.
```
@lyuma Can we check this? I've forgotten about how the mipmap texel squishing works.

BlueCube3310 commented:
I've fixed some bugs with bc6 and added benchmarks - bcdec is faster than both squish and CVTT in all formats. Surprisingly, bcdec's bc6 is 3x faster than CVTT

clayjohn reviewed (approved):
Looks great to me! The performance results are really the cherry on top of the cake!

fire reviewed (approved):
Did not deeply review the code, but the design proposal is good.

akien-mga commented on thirdparty/README.md:
```diff
@@ -650,6 +650,10 @@ comments and a patch is provided in the `patches` folder.
 
+- `bcdec.h`
+  * Upstream: https://github.com/iOrange/bcdec
+  * Version: git (026acf98ea271045cb10713daa96ba98528badb7, 2022)
```
Just noticed this - there's been a handful of commits since (https://github.com/iOrange/bcdec/commits/main/), might be worth updating to the latest, even if there's no functional change. Just so that when I review thirdparty code I know we're up to date.

akien-mga commented:
Thanks!
